joblib/shelf.py

```diff
@@ -0,0 +1,237 @@
+"""
+A context object for temporary shelving arbitrary objects in a data store.
+"""
+
+# Copyright (c) 2018 Inria
+# License: BSD Style, 3 clauses.
+
+import os
+import tempfile
+import weakref
+import shutil
+
+from uuid import uuid4
+
+from .memory import StoreBase
+
+_active_shelf = None
+_active_shelf_mmap = None
+
+
+class JoblibShelfFuture():
+    """A Future-like object for keeping a reference on shelved objects."""
+
+    def __init__(self, data_id, shelf):
+        self._data_id = data_id
+        self._shelf = shelf
+
+    def result(self):
+        """Reloads the data from store and returns it.
+
+        Returns
+        -------
+            The reloaded data.
+        """
+        return self._shelf.store_backend.load_item([self._data_id])
+
+
+###############################################################################
+# class `JoblibShelf`
+###############################################################################
+class JoblibShelf(StoreBase):
+    """A context object for shelving an arbitrary object in a store."""
+
+    # The root of the shelf storage is built with current Python interpreter
+    # process ID
+    _root = 'joblib-shelf-{}'.format(os.getpid())
+    _futures = set()
+
+    def put(self, data):
+        """Put data on the shelf.
+
+        Parameters
+        ----------
+        data: any
+            The data to put on the shelf.
+
+        Returns:
+        --------
+            A future on the shelved data.
+
+        """
+        data_id = uuid4().hex
+        self.store_backend.dump_item([data_id], data)
+
+        future = JoblibShelfFuture(data_id, self)
+
+        def _collect(ref):
+            self._on_collect(data_id)
+        self._futures.add(weakref.ref(future, _collect))
+        return future
+
+    def _on_collect(self, data_id):
+        self.store_backend.clear_location(
+            os.path.join(self.store_backend.location, data_id))
+
+    def __del__(self):
+        # TODO: FIX self.clear()
+        shutil.rmtree(self.store_backend.location)
+
+
+def shelve(input_object):
+    """Shelves an arbitrary object and returns a future on it.
+
+    The input object can then be deleted at any time by the script to save
+    memory. The future, a light-weight object, can be used later to reload the
+    initial object.
+
+    During the life of the future, the input object is kept written on a store
```

Comment:
I would rather say "The input object is kept in a store (by default a file on a disk) as long as the future object exists (technically: as long as there is a reference on the future)".

Reply:
ok